### PR TITLE
[DTensor] Make default RNG semantics match user-passed generator

### PR DESCRIPTION
**Pull request by wconstab**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #160482

Previously, DTensor kept its own copy of the generator state after the
first time a random operator was called on a DTensor. This copy would
evolve independently from the generator outside of DTensor.

After adding support for users to pass a specific generator into
random operators (e.g. `uniform_(..., generator=)`), it was determined
(in discussion on #159991) to change the semantics so that any random
operations performed on DTensor would evolve the state of the publicly
visible generators (either the default one or user-passed one).

The upsides are (1) it is now possible to call torch.manual_seed() at
any point in the program and have a consistent effect on DTensor, (2)
DTensor ops have an observable effect on the generator.  The downside is
that users are now responsible for seeding their generator before using
DTensor, ensuring all ranks use the same seed.

Fixes #159991

confirmed docs rendered OK

image

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @d4l3k @pragupta

### CODE REVIEW

File: test/distributed/tensor/test_random_ops.py [hunk 1]
```diff
@@ -113,22 +113,20 @@ def test_init_with_user_generator(self):
         #     (`torch.distributed.tensor._random._rng_tracker._manual_seed`)
         # (b) If we try to match the semantics of (a) with a user-supplied RNG, they may be very surprised to find that
         #     their RNG object never advances its state after using it with DTensor.
-        # torch.distributed.tensor._random._rng_tracker._manual_seed(55)
-        # rng.manual_seed(55)
-        # torch.nn.init.uniform_(t1, 0.0, 1.0)
-        # torch.nn.init.uniform_(t2, 0.0, 1.0, rng)
-        # self.assertEqual(t1.full_tensor(), t2.full_tensor())
+        torch.manual_seed(55)
```
Comment (bdhirsh) [inline test/distributed/tensor/test_random_ops.py:116]: one maybe-obvious question now that we are requiring the user to provide same-randomness across ranks at the start of their DTensor programs: what is the right way for the user to get that guarantee?

I'm mostly wondering if calling `torch.manual_seed(same_seed_across_ranks)` is enough (assuming they advance RNG in a consistent way across ranks). Or if `manual_seed` gives consistent starting RNG on different machines, or if you always need to broadcast a share seed.

(on typing this out, the "manual_seed gives same starting seed on different hardware" seems like an important property for reproducibility anyway, so I'm guessing this is true? Just checking)

Comment (wconstab) [reply test/distributed/tensor/test_random_ops.py:116]: good question. Right now, you actually do need to do a broadcast to ensure consistency, this PR just proposes moving it outside of Dtensor's internals and making it explicit.

This is how torchtitan initializes its random seeds, for reference:
https://github.com/pytorch/torchtitan/blob/main/torchtitan/distributed/utils.py#L118

I would be interested in brainstorming better approaches here. There are a few pitfalls with the naive solutions that came to mind.
- it's not easy to infer which ranks ought to have the same seed vs different ones.
- DTensor used to always broadcast rank0's seed, but this caused a hang when composing DTensor SPMD parallelisms with Pipeline Parallelism in torchtitan, becuase DTensor just assumed 'the whole world' was SPMD, and it wasn't
- perhaps adding a standalone util, or integrating the util with device-mesh, would let us offer a concise way of expressing which ranks you want seeded which ways and doing it in one shot (albeit, with a collective)

Comment (bdhirsh) [reply test/distributed/tensor/test_random_ops.py:116]: ah thanks for clarifying!

I don't really have any better ideas minus your point about maybe having a helper that we recommend to people. Although - within DTensor, we know which tensors are replicated vs sharded (and therefore which GPUs are supposed to have the same RNG state), so it seems reasonable for having a way to have DTensor (optionally) check if the starting RNG is consistent on the right devices and error otherwise?

Comment (wconstab) [reply test/distributed/tensor/test_random_ops.py:116]: > so it seems reasonable for having a way to have DTensor (optionally) check if the starting RNG is consistent on the right devices and error otherwise?

yes, although, this used to be 'on by default' in dtensor, but that required making the assumption that the 'world group' was SPMD, which we had to disable when supporting pipeline parallelism.

without that assumption, DTensor would still need to be told which groups to check this property over.

>  within DTensor, we know which tensors are replicated vs sharded

well this is not technically true (we may encounter many different tensors with their own placements 'later' after 
initializing the RNG.  However, this did give me the idea that we could try to infer the 'spmd mesh' from the first dtensor we see, since we generally assume there is one spmd mesh.  (which, otoh, is not strictly true either, considering how EP repartitions meshes).

File: docs/source/distributed.tensor.md [hunk 2]
```diff
@@ -179,6 +179,16 @@ specifying the {class}`DeviceMesh` and {class}`Placement` for the {class}`DTenso
 
 ```
 
+### Random Operations
+
+DTensor provides distributed RNG functionality to ensure that random operations on sharded tensors get unique values, and random operations on replicated tensors get the same values. This system requires that all participating
```
Comment (wanchaol) [inline docs/source/distributed.tensor.md:184]: I think maybe mention it is based on the Philox RNG

**Review by wanchaol — APPROVED** [review]
lgtm!

File: test/distributed/tensor/test_random_ops.py [hunk 3]
```diff
@@ -33,6 +33,11 @@
 )
 
 
+def get_generator_seed_for_device_type(device_type: str) -> int:
+    device_module = torch.get_device_module(device_type)
+    return device_module.get_rng_state()[:8].view(torch.int64).item()
```
Comment (zpcore) [inline test/distributed/tensor/test_random_ops.py:38]: Is this a requirement that rng_state is in type of Bytes?

Comment (wconstab) [reply test/distributed/tensor/test_random_ops.py:38]: its a property of all the philox-based generators.  it encodes 2 64bit ints (seed, offset) as a 16-byte tensor. weird..

**Comment by jeffdaily** [comment]
Looks like ciflow/periodic would add the missing rocm tests, but ciflow/trunk should have covered the cuda flows best I can tell.  Not sure for cuda.
